### PR TITLE
Renew membership:  models  (no UI except AppConfig)

### PR DESCRIPTION
**Pull request by weedySeaDragon**
## PT Story:  Renew Membership
- Renewals must also agree to Ethical Guidelines https://www.pivotaltracker.com/story/show/172131253
- Membership Renewal: New doc required every year
https://www.pivotaltracker.com/story/show/154046321

This PR has the changes and additions to **models** for implementing membership renewal.
The only UX is for the new attribute added to the Application Configuration -- so the admin can set/change the value of it.

Once these are ok, then it will be straightforward to add the UX based on this PR.

## Changes proposed in this pull request:
1.  add `membership_expired_grace_period` to AppConfiguration : the number of days after a membership has expired that a user can renew without having to apply as a new member again
2.  track all of the requirements for renewing in the new `RequirementsForRenewal` class.  Also refactor the `RequirementsForMembership` class:  both track all of the non-payment requirements and the payment-related requirements.
3. `ShfApplication:`
    - is an applicant allowed to upload more files to the application?  (Not if it has been accepted or rejected, etc.)
4.  `User` (this is where most additions were made)
    1. file uploads:  get the most recent upload; were any files uploaded during the membership term?
    2. grace period:  is a date within the grace period?  is it after the grace period?
    3. can a user pay the membership fee?
        - is this a renewal?  if so, have the renewal requirements been met?
        - is this a new member? if so, have the membership requirements been met?

---
## Ready for review:
@patmbolger @thesuss @AgileVentures/shf-project-team

### CODE REVIEW

File: app/models/reqs_updaters/requirements_for_renewal.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      membership_guidelines_checklist_done?(user) &&
+      doc_uploaded_during_this_membership_term?(user)
+  end
+
+  # @return [Boolean] - if a user must have a completed Membership Guidelines checklist,
+  #   return true if has been completed  (false if not completed)
+  # else if the user does not have to have a completed Membership Guidelines checklist,
+  #   return true (we assume it's fine)
+  def self.membership_guidelines_checklist_done?(user)
+    UserChecklistManager.completed_membership_guidelines_checklist?(user)
```
Comment (patmbolger) [inline app/models/reqs_updaters/requirements_for_renewal.rb:44]: Does this conform to the logic description ("return true (we assume it's fine)")?

File: app/models/reqs_updaters/requirements_for_renewal.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  def self.doc_uploaded_during_this_membership_term?(user)
+    user.file_uploaded_during_this_membership_term?
+  end
+
+  def self.max_days_can_still_renew
+    AdminOnly::AppConfiguration.config_to_use.membership_expired_grace_period
+  end
+
+  def self.payment_requirements_met?(user)
+    user.payments_current? || user.membership_expired_in_grace_period?
```
Comment (patmbolger) [inline app/models/reqs_updaters/requirements_for_renewal.rb]: Is this a proxy for "... the responsibility of this class to know if the given member has been
 \#   unpaid so long that they cannot simply 'renew' but must re-apply."?  (comment at top of file).

If not, is there any other test to determine that the user has been unpaid too long to simply renew membership?

File: app/models/shf_application.rb [hunk 3]
```diff
@@ -195,6 +198,18 @@ def add_observers
...
+  # @return [boolean] - is the application in a state where the applicant can edit it?
+  def applicant_can_edit?
+    CAN_EDIT_STATES.include? state.to_sym
+  end
+
+
+  # @return [Boolean] - can the applicant upload files for this application?
+  def applicant_can_upload_files?
+    applicant_can_edit? || (accepted? && user.membership_app_and_payments_current?)
+  end
```
Comment (patmbolger) [inline app/models/shf_application.rb]: Since this PR has to do with renewals, and renewals require that a file be uploaded and attached to the _user_, why is this here?

**Review by patmbolger — COMMENTED** [review]
First-pass comments/questions.

File: app/models/reqs_updaters/requirements_for_renewal.rb (hunk 2 above)
Comment (weedySeaDragon) [reply app/models/reqs_updaters/requirements_for_renewal.rb]: No, it's not a proxy. (logic/responsibility for determining that is elsewhere.)

I was thinking that for renewals, a payment is not required if they are still within the grace period.  But perhaps that is not the case.  Deleting it for now.

Good catch.

File: app/models/shf_application.rb (hunk 3 above)
Comment (weedySeaDragon) [reply app/models/shf_application.rb]: Leftover.   Should have been deleted.  Good catch.

File: app/models/user.rb [hunk 4]
```diff
@@ -177,38 +185,91 @@ def membership_current_as_of?(this_date)
     !membership_payment_expire_date.nil? && (membership_payment_expire_date > this_date)
   end
 
+  # The membership term has expired, but are they  still within a 'grace period'?
+  def membership_expired_in_grace_period?(this_date = Time.zone.now)
```
Comment (patmbolger) [inline app/models/user.rb]: Shouldn't the default be ```this_date = Date.current``` ?

File: app/models/user.rb [hunk 5]
```diff
@@ -87,20 +85,38 @@ def self.next_membership_payment_date(user_id)
...
-                              saved_change_to_first_name? ||
-                              saved_change_to_last_name? ||
-                              saved_change_to_membership_number? }
+                 saved_change_to_first_name? ||
+                 saved_change_to_last_name? ||
+                 saved_change_to_membership_number? }
+
+  # @return [ActiveSupport::Duration]
+  def self.membership_expired_grace_period
+    ActiveSupport::Duration.days(AdminOnly::AppConfiguration.config_to_use.membership_expired_grace_period.to_i)
```
Comment (patmbolger) [inline app/models/user.rb]: Couldn't this be simplified to:
```
AdminOnly::AppConfiguration.config_to_use.membership_expired_grace_period.days
```
?

File: app/models/user.rb [hunk 6]
```diff
@@ -87,20 +85,38 @@ def self.next_membership_payment_date(user_id)
...
+  end
+
+  def self.most_recent_upload_method
+    MOST_RECENT_UPLOAD_METHOD
+  end
+
+  # @return [ActiveSupport::Duration]
+  def self.days_can_renew_early
+    ActiveSupport::Duration.days(AdminOnly::AppConfiguration.config_to_use.payment_too_soon_days.to_i)
```
Comment (patmbolger) [inline app/models/user.rb]: Same comment as above.

Comment (patmbolger) [reply app/models/user.rb]: Also, this method (and the similar above, and maybe others here in `User` class) don't seem to be belong here, since they are not attributes of, nor rely on, User class or user instance properties.

Perhaps eventually in a `Membership` class(?) - but for now I'd suggest `AppConfiguration`.

**Review by patmbolger — COMMENTED** [review]
More comments.

File: app/models/user.rb (hunk 4 above)
Comment (weedySeaDragon) [reply app/models/user.rb]: yup.

File: app/models/user.rb (hunk 5 above)
Comment (weedySeaDragon) [reply app/models/user.rb]: doh!  I somehow totally missed that an Integer can be converted to a Duration.
Yes, that will simplify things.

File: app/models/user.rb (hunk 6 above)
Comment (weedySeaDragon) [reply app/models/user.rb]: Agree they ultimately need to be refactored out to a different class.
(I've started work on refactoring.)

They _are_ getting their values from AppConfiguration.

Comment (patmbolger) [reply app/models/user.rb]: @weedySeaDragon I had meant that the method `days_can_renew_early` might be better placed in `AppConfiguration`.  This makes sense, as that class could be imbued with "knowledge" of what the settings represent.

Comment (weedySeaDragon) [reply app/models/user.rb]: The instance method `days_can_renew_early` gets its result from the class method of the same name.  The class method ultimately gets the info from AppConfiguration, which is responsible for providing the definitive values about application settings/configurations.

The class method is the public (external) portal for requesting the info from other sources (classes/instances, whatever).  Since this is a class-level piece of information (e.g. it doesn't vary with instances), I have a User class method responsible for it.

I totally agree that this does not belong in the User class.  But for now, in order to get renewals implemented, I'm putting it here along with all of the other membership term related methods that have piled up in the class.  The User class needs this information in order to do membership term (date) calculations.

We should definitely review what the appropriate responsibility is for the AppConfiguration class, as well as the User class and more.  But I don't want those larger, more involved (and overdue) conversations to derail this right now.

It'd be helpful if you opened a task in Pivotal Tracker to "define the responsibility of the AppConfiguration class" and recorded your thoughts there.  That'd be a good starting place.

**Review by patmbolger — APPROVED** [review]
